crates/ecs/src/lib.rs

```diff
@@ -265,74 +259,293 @@ impl<'systems> Schedule<'systems> for Unordered<'systems> {
+    component_typeid_to_component_vec: HashMap<TypeId, Box<dyn ComponentVec>>,
+    entity_id_to_component_index: HashMap<usize, usize>,
+    last_entity_id_added: usize,
+}
+
+impl Archetype {
+    /// Adds an `entity_id` to keep track of and store components for.
+    fn store_entity(&mut self, entity_id: usize) {
+        let entity_index = self.entity_id_to_component_index.len();
+
+        self.component_typeid_to_component_vec
+            .values_mut()
+            .for_each(|v| v.push_none());
+
+        self.entity_id_to_component_index
+            .insert(entity_id, entity_index);
+        self.last_entity_id_added = entity_id;
+    }
+
+    #[allow(unused)]
+    fn remove_entity(&mut self, entity_id: usize) {
+        if let Some(&index) = self.entity_id_to_component_index.get(&entity_id) {
+            self.component_typeid_to_component_vec
+                .values()
+                .for_each(|vec| vec.swap_remove(index));
+            self.entity_id_to_component_index.remove(&entity_id);
+            // update index of compnonets of entity on last index
+            self.entity_id_to_component_index
+                .insert(self.last_entity_id_added, index);
+        }
+    }
+
+    /// Returns a `ReadComponentVec` with the specified generic type `ComponentType` if it is stored.
+    fn borrow_component_vec<ComponentType: Debug + Send + Sync + 'static>(
+        &self,
+    ) -> ReadComponentVec<ComponentType> {
+        let component_typeid = TypeId::of::<ComponentType>();
+        if let Some(component_vec) = self
+            .component_typeid_to_component_vec
+            .get(&component_typeid)
+        {
+            return borrow_component_vec(component_vec.as_ref());
+        }
+        None
+    }
+
+    /// Returns a `WriteComponentVec` with the specified generic type `ComponentType` if it is stored.
+    fn borrow_component_vec_mut<ComponentType: Debug + Send + Sync + 'static>(
+        &self,
+    ) -> WriteComponentVec<ComponentType> {
+        let component_typeid = TypeId::of::<ComponentType>();
+        if let Some(component_vec) = self
+            .component_typeid_to_component_vec
+            .get(&component_typeid)
+        {
+            return borrow_component_vec_mut(component_vec.as_ref());
+        }
+        None
+    }
+
+    /// Adds a component of type `ComponentType` to the specified `entity`.
+    fn add_component<ComponentType: Debug + Send + Sync + 'static>(
+        &mut self,
+        entity_id: usize,
+        component: ComponentType,
+    ) {
+        if let Some(&entity_index) = self.entity_id_to_component_index.get(&entity_id) {
+            if let Some(mut component_vec) = self.borrow_component_vec_mut::<ComponentType>() {
+                component_vec[entity_index] = Some(component);
+            }
+        }
+    }
+
+    /// Adds a component vec of type `ComponentType` if no such vec already exists.
+    fn add_component_vec<ComponentType: Debug + Send + Sync + 'static>(&mut self) {
```

Comment:
I thought it might be a useful feature for different applications of using Archtypes. Maybe it should be moved to a builder struct. Currently I use it to implement the BigArchtype. I'll add a comment about #72 needing to evaluate if it should stay.

Reply:
👍 

I like the idea of moving it to a builder struct, since then it's clear that you only add new component vecs during a construction-phase, and then once you've built the archetype it is "immutable in it's signature".